src/watch_filter.rs

```diff
@@ -0,0 +1,314 @@
+                .iter()
+                .map(|path| WatchFilter::app_dir_pattern(&app_dir, path))
+                .collect::<Result<_>>()?,
+            ignore_patterns: ignore_patterns
+                .iter()
+                .map(|path| WatchFilter::app_dir_pattern(&app_dir, path))
+                .collect::<Result<_>>()?,
+            process_start_time: SystemTime::now(),
+            files_modified_at: Arc::new(Mutex::new(HashMap::new())),
+        })
+    }
+
+    fn app_dir_pattern(app_dir: &Path, path: &str) -> anyhow::Result<Pattern> {
+        let pat = app_dir.join(path);
+        let pat_str = pat
+            .to_str()
+            .with_context(|| format!("non-unicode pattern {path:?}"))?;
+        Pattern::new(pat_str).with_context(|| format!("invalid glob pattern {path:?}"))
+    }
+
+    fn has_valid_event_kind(&self, event: &Event) -> bool {
+        event.tags.iter().any(|tag| {
+            matches!(
+                tag,
+                Tag::FileEventKind(FileEventKind::Modify(ModifyKind::Data(_)))
+                    | Tag::FileEventKind(FileEventKind::Modify(ModifyKind::Name(_)))
+                    | Tag::FileEventKind(FileEventKind::Create(_))
+                    | Tag::FileEventKind(FileEventKind::Remove(_))
+            )
+        })
+    }
+
+    fn matches_one_of_ignore_patterns(&self, event: &Event) -> bool {
+        event.paths().any(|(path, _)| {
+            self.ignore_patterns
+                .iter()
+                .any(|pattern| pattern.matches_path(path))
+        })
+    }
+
+    fn matches_one_of_path_patterns(&self, event: &Event) -> bool {
+        event.paths().any(|(path, _)| {
+            self.path_patterns
+                .iter()
+                .any(|pattern| pattern.matches_path(path))
+        })
+    }
+
+    /// This is necessary to check due to a bug on macOS emitting modify events on copies
+    /// https://github.com/rust-lang/rust/issues/107130
+    fn path_has_been_actually_modified(&self, event: &Event) -> Result<bool> {
+        // No need to check if a deleted path has been modified because it is gone
+        if event
+            .tags
+            .iter()
+            .any(|tag| matches!(tag, Tag::FileEventKind(FileEventKind::Remove(_))))
+        {
+            return Ok(true);
+        }
+
+        for (path, _) in event.paths() {
+            let metadata = std::fs::metadata(path)?;
+            let path_time = metadata.modified()?;
+            let mut modified_map = self.files_modified_at.lock().unwrap();
+            let path_key = match path.to_str() {
+                Some(s) => s.to_owned(),
+                None => {
+                    tracing::warn!("can't check non-unicode path: {path:?}");
+                    continue;
+                }
+            };
+            let base_time = modified_map
+                .get(&path_key)
+                .unwrap_or(&self.process_start_time);
+            if WatchFilter::one_is_after_another(&path_time, base_time)? {
```

Comment:
It looks like `SystemTime`s can be directly compared with each other (they [implement `PartialOrd`](https://doc.rust-lang.org/std/time/struct.SystemTime.html#impl-PartialOrd%3CSystemTime%3E-for-SystemTime)).

```suggestion
            if path_time > base_time {
```